server/src/main/java/org/elasticsearch/cluster/metadata/DataStreamFailureStore.java

```diff
@@ -0,0 +1,76 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.cluster.metadata;
+
+import org.elasticsearch.cluster.Diff;
+import org.elasticsearch.cluster.SimpleDiffable;
+import org.elasticsearch.common.Strings;
+import org.elasticsearch.common.io.stream.StreamInput;
+import org.elasticsearch.common.io.stream.StreamOutput;
+import org.elasticsearch.xcontent.ConstructingObjectParser;
+import org.elasticsearch.xcontent.ParseField;
+import org.elasticsearch.xcontent.ToXContentObject;
+import org.elasticsearch.xcontent.XContentBuilder;
+import org.elasticsearch.xcontent.XContentParser;
+
+import java.io.IOException;
+
+/**
+ * Holds the data stream failure store metadata that enable or disable the failure store of a data stream. Currently, it
+ * supports the following configurations:
+ * - enabled
+ */
+public record DataStreamFailureStore(boolean enabled) implements SimpleDiffable<DataStreamFailureStore>, ToXContentObject {
```

Comment:
One of the things we're talking about doing is adding a cluster setting that will contain index patterns which will state that failure store is enabled for any matching indices by default. The plan is to fall back to using that cluster setting in the case that failure stores haven't been explicitly enabled or disabled via templates at creation time, or in the state. 

To save on headache going forward, should this boolean be an optional `Boolean`?

Reply:
Good point! thinking further about this, I would expect that the following to be the not explicitly set:

- Explicitly enabled:
```
PUT _data_stream/my-fs/_options
{
   "failure_store": {
     "enabled": true
   }
}
```
- Explicitly disabled:
```
PUT _data_stream/my-fs/_options
{
   "failure_store": {
     "enabled": false
   }
}
```
- No explicit configuration:
```
PUT _data_stream/my-fs/_options
{}
```

This follows the way that the lifecycle works. In template composition, a user can set `failure_store: null` to remove any configuration set by previous templates.

Since this is a new configuration type we could always decide we would like to change the way we work with this, but I thought we should at least consider it.

What do you think?

Reply:
Ah I see, since there is only one property we can just assume the entire section will be removed from options if it's not present. I'm good with that